cmd/ipfs/daemon.go

```diff
@@ -457,6 +469,99 @@ func daemonFunc(req *cmds.Request, re cmds.ResponseEmitter, env cmds.Environment
 	return errs
 }
 
+const mfsRepinInterval = time.Minute
+
+func pinMFSOnChange(cctx *oldcmds.Context, node *core.IpfsNode) (<-chan error, error) {
+	errCh := make(chan error)
+	go func() {
+		var lastRootCid cid.Cid
+		for {
+			// poll periodically
+			tmo := time.NewTimer(mfsRepinInterval)
+			select {
+			case <-cctx.Context().Done():
+				return
+			case <-tmo.C:
+			}
+
+			// get the most recent MFS root cid
+			rootNode, err := node.FilesRoot.GetDirectory().GetNode()
+			if err != nil {
+				select {
+				case errCh <- err:
+				case <-cctx.Context().Done():
+					return
+				}
+				continue
+			}
+			rootCid := rootNode.Cid()
+
+			// do nothing, if MFS has not changed
+			if lastRootCid == rootCid {
+				continue
+			}
+			lastRootCid = rootCid
+
+			// reread the config, which may have changed in the meantime
+			cfg, err := cctx.GetConfig()
+			if err != nil {
+				select {
+				case errCh <- err:
+				case <-cctx.Context().Done():
+					return
+				}
+			}
+
+			// pin on all remote services in parallel to prevent DoS attacks
+			var wg sync.WaitGroup
+			for svcName, svcConfig := range cfg.Pinning.RemoteServices {
+				if svcConfig.Policies.PinMFS == nil || !*svcConfig.Policies.PinMFS {
+					continue
+				}
+				wg.Add(1)
+				go func() {
+					defer wg.Done()
+					pinMFS(cctx.Context(), node, lastRootCid, svcName, svcConfig, errCh)
+				}()
+			}
+			wg.Wait()
+		}
+	}()
+	return errCh, nil
+}
+
+func pinMFS(ctx context.Context, node *core.IpfsNode, cid cid.Cid, svcName string, svcConfig config.RemotePinningService, errCh chan<- error) {
+	c := pinclient.NewClient(svcConfig.Api.ApiEndpoint, svcConfig.Api.ApiKey)
+
+	// Prepare Pin.name
+	opts := []pinclient.AddOption{pinclient.PinOpts.WithName("policy-mfs")}
+
+	// Prepare Pin.origins
+	// Add own multiaddrs to the 'origins' array, so Pinning Service can
+	// use that as a hint and connect back to us (if possible)
+	if node.PeerHost != nil {
+		addrs, err := peer.AddrInfoToP2pAddrs(host.InfoFromHost(node.PeerHost))
+		if err != nil {
+			select {
+			case errCh <- err:
+			case <-ctx.Done():
+			}
+			return
+		}
+		opts = append(opts, pinclient.PinOpts.WithOrigins(addrs...))
+	}
+
+	// Execute remote pin request
+	// TODO: fix panic when pinning service is down
+	_, err := c.Add(ctx, cid, opts...)
```

Comment:
Ideally we'd use the `Replace` function here, but not everyone supports it yet. WDYT @lidel?

----

If we stick with just adding more and more pins and relying on the user's arrangement with the pinning service to remove old copies of MFS then perhaps we want to poll the service on startup for the latest pin and set `lastRootCid`.

Reply:
Keeping old copies pinned and shifting "expiration" burden to pinning service sounds like a feature that some use cases will appreciate, and some will want to avoid. 

It sounds like a flag  that we should add (`KeepOldSnapshots: false` by default)